thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java

```diff
@@ -17,46 +17,98 @@
+ * sliding window. Each sliding window start time and end time is aligned to the data granularity. Each window size is set by the spec.
  */
 public class AnomalyDetectionStageWrapper extends DetectionPipeline {
   private static final String PROP_STAGE_CLASSNAME = "stageClassName";
   private static final String PROP_SPECS = "specs";
   private static final String PROP_METRIC_URN = "metricUrn";
 
-  private final AnomalyDetectionStage anomalyDetectionStage;
+  // moving window detection properties
+  private static final String PROP_MOVING_WINDOW_DETECTION = "movingWindowDetection";
+  private static final String PROP_WINDOW_DELAY = "windowDelay";
+  private static final String PROP_WINDOW_DELAY_UNIT = "windowDelayUnit";
+  private static final String PROP_WINDOW_SIZE = "windowSize";
+  private static final String PROP_WINDOW_UNIT = "windowUnit";
+  private static final String PROP_FREQUENCY = "frequency";
+
+  private static final Logger LOG = LoggerFactory.getLogger(AnomalyDetectionStageWrapper.class);
+
   private final String metricUrn;
+  private final Map<String, Object> specs;
+  private final String stageClassName;
+
+  private final int windowDelay;
+  private final TimeUnit windowDelayUnit;
+  private final int windowSize;
+  private final TimeUnit windowUnit;
+  private final boolean movingWindowDetection;
+  private DatasetConfigDTO dataset;
+  private DateTimeZone dateTimeZone;
+  // need to specify run frequency for minute level detection. Used for moving monitoring window alignment, default to be 15 minutes.
+  private final TimeGranularity functionFrequency;
 
-  public AnomalyDetectionStageWrapper(DataProvider provider, DetectionConfigDTO config, long startTime, long endTime)
-      throws Exception {
+  public AnomalyDetectionStageWrapper(DataProvider provider, DetectionConfigDTO config, long startTime, long endTime) {
     super(provider, config, startTime, endTime);
 
     Map<String, Object> properties = config.getProperties();
     Preconditions.checkArgument(properties.containsKey(PROP_STAGE_CLASSNAME), "Missing " + PROP_STAGE_CLASSNAME);
 
-    this.anomalyDetectionStage = loadAnomalyDetectorStage(MapUtils.getString(properties, PROP_STAGE_CLASSNAME));
-    Map<String, Object> specs = MapUtils.getMap(properties, PROP_SPECS);
+    this.specs = MapUtils.getMap(properties, PROP_SPECS);
+    this.stageClassName = MapUtils.getString(properties, PROP_STAGE_CLASSNAME);
     this.metricUrn = MapUtils.getString(config.getProperties(), PROP_METRIC_URN);
-    if (this.metricUrn != null){
-      specs.put(PROP_METRIC_URN, metricUrn);
+    if (this.metricUrn != null) {
+      this.specs.put(PROP_METRIC_URN, metricUrn);
     }
-    this.anomalyDetectionStage.init(specs, config.getId(), startTime, endTime);
+
+    this.movingWindowDetection = MapUtils.getBooleanValue(this.specs, PROP_MOVING_WINDOW_DETECTION, false);
+    this.windowDelay = MapUtils.getIntValue(this.specs, PROP_WINDOW_DELAY, 0);
```

Comment:
Can you include some comments in the code on what and why we need this windowDelay for?

Reply:
added